src/serialize/print_driver.cc

```diff
@@ -0,0 +1,87 @@
+#include <config.h>
+#include <cstring>
+#include <serialize/print_driver.h>
+
+namespace freetensor {
+
+std::string dumpTarget(const Ref<Target> &target_) {
+
+    auto target = target_.isValid() ? target_ : Config::defaultTarget();
+
+    std::string ret =
+        target->toString() + " " + std::to_string(target->useNativeArch());
+
+    switch (target->type()) {
+    case TargetType::GPU: {
+        auto _target = target.as<GPU>();
+        auto _computeCapability = _target->computeCapability();
+        if (_computeCapability.isValid()) {
+            ret += " : " + std::to_string(_computeCapability->first) + " " +
+                   std::to_string(_computeCapability->second);
+        } else {
+            ret += " ;";
+        }
+        break;
+    }
+    case TargetType::CPU:
+        break;
+
+    default:
+        ASSERT(false);
+    }
+
+    return ret;
+}
+
+std::string dumpDevice(const Ref<Device> &device_) {
+
+    auto device = device_.isValid() ? device_ : Config::defaultDevice();
+
+    std::string ret = "DEV " + std::to_string(device->num()) + " " +
+                      dumpTarget(device->target());
+    return ret;
+}
+
+std::string dumpArray(const Ref<Array> &array_) {
+
+    /**
+     * The string is constructed as follow (Separated by space):
+     *
+     * {"ARR"} +
+     * {dtype} +
+     * {shape.size} + {shape[0]} + ... + {shape[shape.size - 1]} +
+     * {ptrs.size} + {ptrs[0].dev_"#"} + ... + {ptrs[ptrs.size - 1].dev_"#"} +
+     * {Arraydata (uint8_t, strlen = array_->size(), e.g.!@#$%^&*1)}
+     *
+     * Note: Arraydata may have invisible characters
+     */
```

Comment:
You are mixing character strings with byte strings. If it is a character string, it should be stored in a `std::string`. If it is a byte string, it cannot, because there may be bytes corresponding to illegal characters, e.g. `\0`.

In RPC, byte strings are handled in a designated type (Base64 for XMLRPC). You have two options:

1. Separate the meta data (types, sizes) and data bytes.
2. Encode the meta data in a binary form together with the data bytes.